File: limablu_mobile/lib/services/prefs.dart
// lib/services/prefs.dart
import 'package:shared_preferences/shared_preferences.dart';

class AppPrefs {
  static const _kOrderKey = 'deduction_order';
  late final SharedPreferences _sp;

  Future<void> init() async {
    _sp = await SharedPreferences.getInstance();
  }

  Future<List<String>> getDeductionOrder() async {
    final v = _sp.getStringList(_kOrderKey);
    return v ?? const ['allowance', 'tokens', 'postpaid'];
  }

  Future<void> setDeductionOrder(List<String> order) async {
    await _sp.setStringList(_kOrderKey, order);
  }

  /// Rotate current order forward, e.g. [allowance, tokens, postpaid] ->
  /// [tokens, postpaid, allowance]. Returns the new order.
  Future<List<String>> rotateDeductionOrder() async {
    final cur = await getDeductionOrder();
    if (cur.isEmpty) return cur;
    final next = [...cur.skip(1), cur.first];
    await setDeductionOrder(next);
    return next;
  }
}

File: emappa-mobile/src/services/prefs.ts
import AsyncStorage from '@react-native-async-storage/async-storage';

const STORAGE_KEY = 'deduction_order';

class AppPrefs {
  async getDeductionOrder(): Promise<string[]> {
    try {
      const raw = await AsyncStorage.getItem(STORAGE_KEY);
      if (raw) {
        return JSON.parse(raw) as string[];
      }
    } catch (e) {
      console.error('Failed to get deduction order:', e);
    }
    return ['allowance', 'tokens', 'postpaid'];
  }

  async setDeductionOrder(order: string[]): Promise<void> {
    try {
      await AsyncStorage.setItem(STORAGE_KEY, JSON.stringify(order));
    } catch (e) {
      console.error('Failed to set deduction order:', e);
    }
  }

  // Rotate current order forward
  // e.g. [allowance, tokens, postpaid] -> [tokens, postpaid, allowance]
  async rotateDeductionOrder(): Promise<string[]> {
    const current = await this.getDeductionOrder();
    if (current.length === 0) return current;

    const next = [...current.slice(1), current[0]];
    await this.setDeductionOrder(next);
    return next;
  }
}

// Singleton instance
export const appPrefs = new AppPrefs();
